File: bukkit/src/main/java/io/wdsj/asw/bukkit/listener/SignListener.java
package io.wdsj.asw.bukkit.listener;

import com.github.houbb.heaven.util.lang.StringUtil;
import io.wdsj.asw.bukkit.AdvancedSensitiveWords;
import io.wdsj.asw.bukkit.event.ASWFilterEvent;
import io.wdsj.asw.bukkit.event.EventType;
import io.wdsj.asw.bukkit.impl.list.AdvancedList;
import io.wdsj.asw.bukkit.manage.notice.Notifier;
import io.wdsj.asw.bukkit.manage.permission.Permissions;
import io.wdsj.asw.bukkit.manage.punish.Punishment;
import io.wdsj.asw.bukkit.proxy.bungee.BungeeSender;
import io.wdsj.asw.bukkit.proxy.velocity.VelocitySender;
import io.wdsj.asw.bukkit.setting.PluginMessages;
import io.wdsj.asw.bukkit.setting.PluginSettings;
import io.wdsj.asw.bukkit.util.Utils;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.SignChangeEvent;

import java.util.List;

import static io.wdsj.asw.bukkit.AdvancedSensitiveWords.*;
import static io.wdsj.asw.bukkit.util.TimingUtils.addProcessStatistic;
import static io.wdsj.asw.bukkit.util.Utils.messagesFilteredNum;

public class SignListener implements Listener {
    private String outMessage = "";
    private List<String> outList = new AdvancedList<>();
    private String outProcessedMessage = "";


    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOW)
    public void onSign(SignChangeEvent event) {
        if (!isInitialized) return;
        if (!settingsManager.getProperty(PluginSettings.ENABLE_SIGN_EDIT_CHECK)) return;
        Player player = event.getPlayer();
        if (player.hasPermission(Permissions.BYPASS)) return;
        boolean shouldSendMessage = false;
        long startTime = System.currentTimeMillis();
        List<Integer> indexList = new AdvancedList<>();
        StringBuilder originalMultiMessages = new StringBuilder();
        for (int line = 0; line < 4; ++line) {
            String originalMessage = event.getLine(line);
            if (settingsManager.getProperty(PluginSettings.PRE_PROCESS) && originalMessage != null) originalMessage = originalMessage.replaceAll(Utils.getPreProcessRegex(), "");
            assert originalMessage != null;
            List<String> censoredWordList = AdvancedSensitiveWords.sensitiveWordBs.findAll(originalMessage);
            if (!censoredWordList.isEmpty()) {
                String processedMessage = AdvancedSensitiveWords.sensitiveWordBs.replace(originalMessage);
                outMessage = originalMessage;
                outProcessedMessage = processedMessage;
                outList = censoredWordList;
                if (settingsManager.getProperty(PluginSettings.SIGN_METHOD).equalsIgnoreCase("cancel")) {
                    shouldSendMessage = true;
                    event.setCancelled(true);
                    break;
                }
                event.setLine(line, processedMessage);
                shouldSendMessage = true;
            } else if (StringUtil.isNotEmptyTrim(originalMessage)) {
                indexList.add(line);
                originalMultiMessages.append(originalMessage);
            }
        }
        if (settingsManager.getProperty(PluginSettings.SIGN_MULTI_LINE_CHECK) && !indexList.isEmpty()) {
            String originalMessagesString = originalMultiMessages.toString();
            List<String> censoredWordList = AdvancedSensitiveWords.sensitiveWordBs.findAll(originalMessagesString);
            if (!censoredWordList.isEmpty()) {
                String processedMessagesString = AdvancedSensitiveWords.sensitiveWordBs.replace(originalMessagesString);
                outMessage = originalMessagesString;
                outProcessedMessage = processedMessagesString;
                outList = censoredWordList;
                if (settingsManager.getProperty(PluginSettings.SIGN_METHOD).equalsIgnoreCase("cancel")) {
                    shouldSendMessage = true;
                    event.setCancelled(true);
                } else {
                    shouldSendMessage = true;
                    for (int i : indexList) {
                        event.setLine(i, processedMessagesString);
                    }
                }
            }
        }

        if (shouldSendMessage && settingsManager.getProperty(PluginSettings.ENABLE_API)) {
            Bukkit.getPluginManager().callEvent(new ASWFilterEvent(player, outMessage, outProcessedMessage, outList, EventType.SIGN, false));
        }

        if (settingsManager.getProperty(PluginSettings.SIGN_SEND_MESSAGE) && shouldSendMessage) {
            player.sendMessage(ChatColor.translateAlternateColorCodes('&', messagesManager.getProperty(PluginMessages.MESSAGE_ON_SIGN)));
        }

        if (settingsManager.getProperty(PluginSettings.LOG_VIOLATION) && shouldSendMessage) {
            Utils.logViolation(player.getName() + "(IP: " + Utils.getPlayerIp(player) + ")(Sign)", outMessage + outList);
        }

        if (settingsManager.getProperty(PluginSettings.HOOK_VELOCITY) && shouldSendMessage) {
            VelocitySender.send(player, EventType.SIGN, outMessage);
        }

        if (settingsManager.getProperty(PluginSettings.HOOK_BUNGEECORD)) {
            BungeeSender.send(player, EventType.SIGN, outMessage);
        }

        if (settingsManager.getProperty(PluginSettings.ENABLE_DATABASE)) {
            databaseManager.checkAndUpdatePlayer(player.getName());
        }

        if (shouldSendMessage) {
            messagesFilteredNum.getAndIncrement();
            long endTime = System.currentTimeMillis();
            addProcessStatistic(endTime, startTime);
            if (settingsManager.getProperty(PluginSettings.NOTICE_OPERATOR)) Notifier.notice(player, EventType.SIGN, outMessage);
            if (settingsManager.getProperty(PluginSettings.SIGN_PUNISH)) Punishment.punish(player);
        }
    }
}

File: bukkit/src/main/kotlin/io/wdsj/asw/bukkit/listener/SignListener.kt
package io.wdsj.asw.bukkit.listener

import io.wdsj.asw.bukkit.AdvancedSensitiveWords.sensitiveWordBs
import io.wdsj.asw.bukkit.AdvancedSensitiveWords.settingsManager
import io.wdsj.asw.bukkit.setting.PluginMessages
import io.wdsj.asw.bukkit.setting.PluginSettings
import io.wdsj.asw.bukkit.type.ModuleType
import io.wdsj.asw.bukkit.integration.sign.SignFakeViewService
import io.wdsj.asw.bukkit.util.PlayerProcessingGuard
import io.wdsj.asw.bukkit.util.Utils
import io.wdsj.asw.bukkit.util.ViolationReporter
import io.wdsj.asw.bukkit.util.context.SignContext
import io.wdsj.asw.bukkit.util.message.MessageUtils
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.block.SignChangeEvent

// TODO: Paper event handler
class SignListener : Listener {
    private val processingGuard = PlayerProcessingGuard()
    private val violationReporter = ViolationReporter()

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onSign(event: SignChangeEvent) {
        if (!settingsManager.getProperty(PluginSettings.ENABLE_SIGN_EDIT_CHECK)) return
        if (event.lines().isEmpty()) return

        val player = event.player
        if (processingGuard.shouldSkipBasic(player)) return

        val startTime = System.currentTimeMillis()
        val attemptedLines = event.lines().toList()
        val lineScan = censorSingleLines(event)
        val violation = lineScan.violation
            ?: censorMultiLine(event, lineScan)
            ?: censorContext(event, player)
            ?: return

        if (isCancelMode() && settingsManager.getProperty(PluginSettings.SIGN_FAKE_ON_CANCEL)) {
            SignFakeViewService.recordCancelledEdit(
                event,
                player,
                attemptedLines,
                violation.content,
                violation.censoredWords,
            )
        }

        if (settingsManager.getProperty(PluginSettings.SIGN_SEND_MESSAGE)) {
            MessageUtils.sendMessage(player, PluginMessages.MESSAGE_ON_SIGN)
        }

        val location = event.block.location
        val locationLog = "World: ${location.world?.name ?: "Unknown"}, X: ${location.x}, Y: ${location.y}, Z: ${location.z}"
        violationReporter.reportWithCustomLogPrefix(
            player = player,
            moduleType = ModuleType.SIGN,
            content = violation.content,
            censoredWords = violation.censoredWords,
            logPrefix = "${player.name}(IP: ${Utils.getPlayerIp(player)})(Sign)($locationLog)",
            startTime = startTime,
            punish = settingsManager.getProperty(PluginSettings.SIGN_PUNISH),
        )
    }

    private fun censorSingleLines(event: SignChangeEvent): SignLineScan {
        var violation: SignViolation? = null
        val cleanLineIndexes = mutableListOf<Int>()
        val cleanLineContent = StringBuilder()

        for (lineIndex in event.lines().indices) {
            val originalComponent = event.line(lineIndex) ?: continue
            val originalMessage = preprocess(MessageUtils.plainText(originalComponent))
            val censoredWords = sensitiveWordBs.findAll(originalMessage)

            if (censoredWords.isEmpty()) {
                if (originalMessage.trim().isNotEmpty()) {
                    cleanLineIndexes.add(lineIndex)
                    cleanLineContent.append(originalMessage)
                }
                continue
            }

            violation = SignViolation(originalMessage, censoredWords)
            if (isCancelMode()) {
                event.isCancelled = true
                continue
            }
            val processedMessage = sensitiveWordBs.replace(originalMessage)
            event.line(lineIndex, MessageUtils.replaceLiteral(originalComponent, originalMessage, processedMessage))
        }

        return SignLineScan(violation, cleanLineIndexes, cleanLineContent.toString())
    }

    private fun censorMultiLine(event: SignChangeEvent, lineScan: SignLineScan): SignViolation? {
        if (!settingsManager.getProperty(PluginSettings.SIGN_MULTI_LINE_CHECK)) return null
        if (lineScan.cleanLineIndexes.isEmpty()) return null

        val censoredWords = sensitiveWordBs.findAll(lineScan.cleanLineContent)
        if (censoredWords.isEmpty()) return null

        if (isCancelMode()) {
            event.isCancelled = true
        } else {
            val processedMessage = sensitiveWordBs.replace(lineScan.cleanLineContent)
            for (lineIndex in lineScan.cleanLineIndexes) {
                event.line(lineIndex, MessageUtils.plainTextComponent(processedMessage))
            }
        }

        return SignViolation(lineScan.cleanLineContent, censoredWords)
    }

    private fun censorContext(event: SignChangeEvent, player: Player): SignViolation? {
        if (!settingsManager.getProperty(PluginSettings.SIGN_CONTEXT_CHECK)) return null

        val originalAllMessage = event.lines().joinToString("") { MessageUtils.plainText(it) }
        SignContext.addMessage(player, originalAllMessage)
        val originalContext = SignContext.getHistory(player).joinToString("")
        val censoredWords = sensitiveWordBs.findAll(originalContext)
        if (censoredWords.isEmpty()) return null

        SignContext.pollPlayerContext(player)
        event.isCancelled = true
        return SignViolation(originalContext, censoredWords)
    }

    private fun preprocess(text: String): String {
        if (!settingsManager.getProperty(PluginSettings.PRE_PROCESS)) return text
        return text.replace(Utils.preProcessRegex.toRegex(), "")
    }

    private fun isCancelMode(): Boolean {
        return settingsManager.getProperty(PluginSettings.SIGN_METHOD).equals("cancel", ignoreCase = true)
    }

    private data class SignLineScan(
        val violation: SignViolation?,
        val cleanLineIndexes: List<Int>,
        val cleanLineContent: String,
    )

    private data class SignViolation(
        val content: String,
        val censoredWords: List<String>,
    )
}
